Introduce missing team role if necessary

raboof opened the pull request:
When receiving gossip from a node that did not contain any team
information (such as gossip from a node running a previous version of
Akka), add the default team role during deserialization.

Refs #23243

johanandren commented on akka-bench-jmh/src/main/scala/akka/cluster/ClusterMessageSerializerBenchmark.scala:
```diff
@@ -0,0 +1,35 @@
+  @TearDown
+  def tearDown: Unit =
+    system.terminate()
+}
```
On purpose in this PR?

raboof replied:
Hmm this doesn't really need to be saved for posterity I guess :)

johanandren replied:
I don't mind if it is, was just curious

johanandren commented on akka-cluster/src/main/scala/akka/cluster/protobuf/ClusterMessageSerializer.scala:
```diff
@@ -336,7 +336,15 @@ class ClusterMessageSerializer(val system: ExtendedActorSystem) extends BaseSeri
+      roles.find(_.startsWith("team-")) match {
+        case Some(_) ⇒ roles
+        case None    ⇒ roles + "team-default"
+      }
```
How about an imperative loop through with the `Set` builder and keep a boolean for seen rather than two loops through the set? (Also, use the constants for the prefix and default from `ClusterSettings`)

raboof replied:
Agreed on using the constant, that wasn't on the `wip-multi-dc-dev` branch yet though I believe :).

Not sure if a loop will be much nicer here, after all we don't really expect many roles. Will give it a shot though.

johanandren replied:
I would be nicer because it doesn't go through the original java-list twice and doesn't allocate a lambda, so only performance related.

raboof replied:
wrote it out, it's now faster than the original implementation that didn't add a default team :)

patriknw commented on akka-cluster/src/main/scala/akka/cluster/protobuf/ClusterMessageSerializer.scala:
```diff
@@ -336,7 +336,15 @@ class ClusterMessageSerializer(val system: ExtendedActorSystem) extends BaseSeri
-        member.getRolesIndexesList.asScala.map(roleMapping(_))(breakOut))
+        rolesFromProto(member.getRolesIndexesList.asScala))
+
+    def rolesFromProto(roleIndexes: Seq[Integer], resultSoFar: Set[String] = Set.empty): Set[String] = {
```
resultSoFar not used

patriknw commented on akka-cluster/src/main/scala/akka/cluster/protobuf/ClusterMessageSerializer.scala:
```diff
@@ -336,7 +336,15 @@ class ClusterMessageSerializer(val system: ExtendedActorSystem) extends BaseSeri
+      val roles = roleIndexes.map(roleMapping(_)).toSet
+      roles.find(_.startsWith("team-")) match {
+        case Some(_) ⇒ roles
+        case None    ⇒ roles + "team-default"
```
use the new constants in `ClusterSettings`

ktoso commented on akka-bench-jmh/src/main/scala/akka/cluster/ClusterMessageSerializerBenchmark.scala:
```diff
@@ -0,0 +1,35 @@
+
+  @TearDown
+  def tearDown: Unit =
+    system.terminate()
```
await that termination finishes - this will leak an alive system into the time when a new benchmark is already running and may impact results

patriknw replied:
it's good to do such cleanup, but in practice I think it's always forked

ktoso replied:
Still bad - uses resources / could still be doing stuff which can have impact on the benchmark's next run

ktoso reviewed (approved):
LGTM after await added

raboof committed:
More efficient `rolesFromProto`

Now actually outperforms the previous implementation. Still room for
improvement as this probably checks for duplicates in the set on each add,
but creating our own array-backed set here is probably going overboard :).

raboof committed:
Introduce missing team role if necessary (#23243)

When receiving gossip from a node that did not contain any team
information (such as gossip from a node running a previous version of
Akka), add the default team role during deserialization.

raboof committed:
Simpler implementation of adding default role

raboof committed:
Fixes following rebase
